### ISSUE TITLE
AI day2

### ISSUE THREAD

**Issue by Romangduck** [Romangduck/Day-1-Linux#11]
﻿
0.활성화함수

활성화 함수(Activation Function)란 인공 신경망(Neural Network)에서 사용되는 은닉층(Hidden Layer)을 활성화하기 위한 함수입니다.

순전파(Forward Propagation) & 역전파(Back Propagation)의 풀이에서 처럼 가중치(Weight)와 편향(Bias)을 갱신(Update)하기 위해 사용되는 주요한 함수입니다.

시스템(System)에 포함된 노드(Node)는 출력값에 동일한 영향을 미치지 않습니다. 즉, 각 노드마다 전달되어야 하는 정보량이 다르다는 의미가 됩니다.

예를 들어 나의 출근 시간을 예측하기 위한 모델(Model)을 구현한다고 가정해보겠습니다.

입력값이 일어난 시간(x1), 기상 상태(x2)라면, 기상 상태(x2)값 보다는 일어난 시간(x1)이 더 영향력이 크다는 것을 직관적으로 알 수 있습니다.

그러므로 연산 과정에서 일어난 시간(x1)은 더 많이 활성화(Activate)되어야 하며, 기상 상태(x2)는 비교적 비활성화(Deactivate) 되어야 합니다.

활성화 함수(Activation Function)는 비선형(Non-linear) 구조를 가져 역전파 과정에서 미분값을 통해 학습이 진행될 수 있게 합니다.

만약, 활성화 함수가 선형(Linear) 구조라면, 미분 과정에서 항상 상수가 나오게 되므로 학습이 진행되지 않습니다.

선형 함수가 아니라 비선형 함수를 사용하는 이유는 딥러닝 모델의 레이어 층을 깊게 가져갈 수 있기 때문이다.

Sigmoid : 이진분류에서 무조건 사용됨, 출력층 레이어의 활성화 함수를 sigmoid로 해줘야함

ex).

model.add(Dense(1,activation="sigmoid"))

relu : 은닉층에 사용 되는 활성화 함수, 성능이 향상됨

가중치그래프의 복잡성을 늘려주는 역할을 한다.

분류문제에서라면 좀 더 쉽게 이해할 수 있는데 그냥 직선으로 나누는 것 보다 데이터에 맞춰 선이 구부러질 수록 더 잘 나눌 수 있게 된다. 해당 데이터에 대해서 만큼은 더 잘 분류하는 모델이 되는 셈이다.

대신 이 비선형성의 증가는 한가지 리스크를 가지고 있는데 바로 오버피팅, 과적합이다.

그래프가 복잡해지면서 더욱 훈련데이터에 대해 철저해질 수록 다른 데이터가 들어왔을 때는 오히려 제대로 구분을 못하는 현상이 발생할 수 있기 때문이다.

정리하자면

1) relu를 많이 쓰면 비선형성이 는다 -> 음수부분 비선형화

2) 비선형성이 늘어나면 (모델이) 좋아진다 -> 좀 더 복잡한 그래프 생성, but 과적합의 위험성↑

model.add(Dense(100,activation="relu"))

model.add(Dense(80,activation="relu"))

model.add(Dense(100,activation="relu"))

linear :

linear 함수는 입력값에 대해 선형 연산을 수행하여 결과값을 출력하는 함수입니다. 입력과 가중치의 곱과 편향(bias)을 더한 값이 출력값으로 나오는 단순한 함수입니다.

선형함수는 다음과 같이 수식으로 나타낼 수 있습니다.

y = wx + b

w는 가중치(weight)를 의미

b는 편향(bias)을 의미

주로 입력층에 사용

model.add(Dense(124,activation='linear',input_dim=13))

1. train_test_split

Train Set

모델의 학습만을 위해서 사용

parameter나 feature 등을 수정하여 모델의 성능을 높이는 작업에 사용

Test Set

최종적으로 모델의 성능을 평가

실사용 되었을 때 모델이 얼마나 좋은 성능을 발휘 할 수 있을지 알아보는 것

Validation Set

모델의 학습에 직접적으로 관여하지 않음 , 학습이 끝난 모델에 적용

최종적으로 모델을 fine tuning 하는데에 사용

validation split은 데이터를 훈련 세트와 검증 세트로 나눌 때 사용되는 방법 중 하나입니다. 훈련 세트에서 일부 데이터를 검증 세트로 사용하여 모델을 훈련하고 검증 세트에서 모델을 평가합니다. 이 방법은 테스트 세트가 없는 상황에서 모델을 평가하기 위해 사용됩니다.

보통 validation split은 Keras와 같은 딥러닝 프레임워크에서 지원됩니다. Keras에서 fit() 함수를 호출할 때 validation_split 인자를 지정하여 데이터를 훈련 세트와 검증 세트로 자동으로 나눌 수 있습니다. 예를 들어,

validation_split=0.2라고 지정하면 데이터의 20%가 검증 세트로 사용됩니다

일반적으로 Train의 size를 0.7 (70%) , Test의 size를 0.3 (30%)로 잡아주는게 좋음

7 : 3 이 적합한 Training : Test 비율이다. 8~9: 2 ~1 로 넘어가면 과적합이 날 수 있음

6 : 2 : 2

Training : Validation : Test

2. matplotli b – scatter (산점도 )

scatter() 함수는 데이터를 이용하여 산점도를 그려줍니다. plt.show() 함수를 호출하여 그래프를 출력할 수 있습니다.

출력된 그래프는 x축과 y축을 가지고 있으며, 각 점은 x와 y의 값에 대응합니다. 이렇게 산점도를 그리면 데이터 간의 관계를 직관적으로 파악할 수 있습니다.

######## scatter 시각화

import matplotlib.pyplot as plt

plt.scatter(x,y) # 산점도 그리기

plt.plot(x,y_predict,color='blue')

plt.show()

대표사진 삭제
사진 설명을 입력하세요.

3. R2 score ( 결정계수 )

R2 score는 회귀 모델의 성능을 평가하는 지표 중 하나입니다. R2 score는 모델이 실제 값과 얼마나 잘 일치하는지를 나타내는 값으로, 0부터 1까지의 범위를 가지며 1에 가까울수록 모델의 예측이 실제 값과 일치한다는 것을 의미합니다.

R2 score는 0 이상의 값이며, 1에 가까울수록 모델이 실제 값과 잘 일치한다는 것을 의미합니다. 반면에, R2 score가 음수가 나올 경우 모델의 예측이 실제 값보다 나쁘다는 것을 의미합니다.

### R2score

from sklearn.metrics import r2_score, accuracy_score

r2 = r2_score(y, y_predict)

print('r2스코어 : ', r2)

회귀분석할때 사용됨

4. 회귀분석

연속된 값을 예측

과거의 주가 데이터를 가지고 미래 주가를 예측하거나, 자동차 배기량이나 연식 등 중고차 정보를 이용하여

가격을 예측

회귀 분석(Regression analysis)은 독립 변수(X)와 종속 변수(Y) 간의 관계를 모델링하는 기법입니다. 이를 통해 독립 변수가 종속 변수에 어떤 영향을 미치는지 파악하고, 이를 바탕으로 미래 값을 예측할 수 있습니다.

기존 A의 값을 바탕으로 비슷한 종류 B의 값을 예측하는것.

회귀모델에서 인풋과 아웃풋 활성화 함수는 linear

r2 를 사용함 -> 모델의 성능을 평가하는 지표

5. 이진분류

이진분류의 경우 참 또는 거짓을 판별하기 때문에 출력 값이 하나.

둘 중 하나의 값을 갖는 출력값을 가지고 입력 데이트를 분류.

출력 값을 sigmoid 함수를 이용하여 0과 1로 가공

로지스틱 회귀는 이진분류모델을 분석하기 좋은 모델 ; 직선보다 곡선을 통해 분류를함

즉, 예측하고자 하는 대상이 두 가지 중 하나인 경우를 말합니다. 예를 들어, 이메일이 스팸인지 아닌지를 분류하는 문제나, 환자가 양성인지 음성인지를 예측하는 문제가 이진 분류 문제입니다.

이진분류에서는 컴파일 단계에서, binary_crossentropy를 사용해야함.

6. 다중분류

다중 분류 모델은 타깃의 종류가 여러 개이기 때문에 출력 값도 여러 개

다중 분류는 softmax 함수를 사용하여 0과 1사이의 값으로 가공

입력 데이터가 주어졌을 때, 3개 이상의 클래스 중 하나로 분류하는 문제입니다. 예를 들어, 이미지가 주어졌을 때 그 이미지가 고양이, 개, 혹은 새 중 어느 것에 해당하는지 분류하는 문제 등이 있습니다.

다중 분류 모델에서는 one-hot encoding을 사용/

7. 원 핫 인코딩 (One Hot Encoding)

원핫인코딩(One-Hot Encoding)은 범주형 데이터를 수치형 데이터로 변환하는 방법 중 하나입니다. 원핫인코딩을 적용하면 각 범주형 데이터를 벡터로 표현할 수 있습니다. 예를 들어, 세 가지 종류의 과일을 분류하는 모델을 만들기 위해 사과, 바나나, 오렌지를 범주형 데이터로 사용한다면, 사과를 1 0 0, 바나나를 0 1 0, 오렌지를 0 0 1로 표현할 수 있습니다. 이렇게 범주형 데이터를 벡터로 표현하면 모델이 이해하기 쉬워지고, 성능을 높일 수 있습니다.

### one hot encoding

from keras.utils import to_categorical

y = to_categorical(y)

print(y)

컴파일 단계에서 categorical_crossentropy를 사용해야함.

*sparse_categorical_crossentropy를 사양한다면 할 필요없음

8. 분류분석

분류 분석(Classification Analysis)은 주어진 데이터를 미리 정의된 여러 개의 클래스(Class) 중 하나로 분류하는 것을 목적으로 하는 분석 방법입니다. 분류 분석은 지도학습(Supervised Learning)의 일종으로, 주어진 데이터와 해당 데이터의 클래스 정보를 이용하여 모델을 학습하고, 학습된 모델을 이용하여 새로운 데이터가 어떤 클래스에 속하는지 분류합니다.

)_있는 정보 분류함

학생들의 국어, 영어, 수학 점수를 통해 내신등급을 알아맞히는 것

카드회사에서 회원들의 가입 정보를 통해 1년후 신용등급을 알아맞히는 것

accuracy가 분류 분석할때 사용됨

9. 난수값 (random_state)

random_state 는 데이터를 난수값에 의해 추출한다는 의미이며, 중요한 파라미터이다.

1~ 100까지 일련번호로 된 100개의 데이터를 train_test_split(.., test_size=0.2) 로 수행하면 해당 함수를 첫번째 수행할 때는 1~80 번이 train, 81~100번이 test가 될 수 있지만, 다시 수행하면 이번에 21~100번이 train, 1~20번이 test가 될 수 있습니다. 80%, 20% 로 나누는건 동일하지만 함수를 수행 시마다 추출한 레코드들을 달라질수 있습니다. 내부적으로 80%, 20% 로 나눌때 random 함수를 적용합니다.

random_state=1 이라고 하면 바로 이 random 함수의 seed 값을 고정시키기 때문에 여러번 수행하더라도 같은 레코드를 추출합니다.

ex). x_train, x_test, y_train,y_test = train_test_split(

x, y, test_size=0.3,train_size=0.7, random_state=72, shuffle=True

여기서 shuffle=True는 데이터를 섞어서 가지고 올 것인지를 정하는것이다.

-----------------------------------------------------------------------------------------------------------------------------

print(datasets.DESCR)

print(datasets.feature_names)

첫번째 결과값 -> datasets의 설명 , 속성정보와

input_dim 의 값을 확인 할 수있다.

두번째 결과값 -> 특징들을 나열

x = datasets.data #(178,13)

y = datasets.target #(178,)

print(x.shape,y.shape)

x_train,x_test,y_train,y_test = train_test_split(

x,y,train_size=0.7,random_state=100,shuffle=True

)

print(x_train.shape,y_train.shape) #(124,13) (124,)

print(x_test.shape,y_test.shape) #(54,13) (54,)

print(y_test) # 클래스 ; 모델구성에서 출력층 노드값 참고사항

위 코드는 scikit-learn의 load_wine() 함수를 사용하여 와인 데이터를 로드하고, 이를 train_test_split() 함수를 사용하여 70%의 데이터를 train set으로, 30%의 데이터를 test set으로 분리한 후, test set의 레이블을 출력하는 코드입니다.

y_test를 출력한 이유는 모델의 출력층에서 몇 개의 노드를 사용해야 하는지 알기 위해서입니다. wine 데이터셋은 3개의 클래스로 분류하는 문제이므로, 출력층의 노드 수는 3개여야 합니다. 따라서 y_test를 출력하여 클래스 수를 알아낼 수 있습니다.

즉 ; 마지막 줄을통해 모델구성 단계에서 출력층의 노드값을 확인할 수 있음.

Accuracy)

실제값과 예측값을 비교하여 두 값의 차이를 구하는것 -

실제값 - 예측값 = 0 이라면 오차가 없는것 ( 정답률 100 % )

accuracy의 값이 0 이라면 그냥 망한거임. 답이없음.

accuracy 값이 높다고 해서 반드시 오차가 없는 것은 아닙니다. accuracy는 모델이 예측한 정확한 예측값(참 양성과 참 음성)의 비율을 나타내는 지표이기 때문에 모델이 예측한 예측값 중에서 실제 값과 일치하지 않는 값(거짓 양성과 거짓 음성)이 있을 수 있습니다. 따라서 accuracy 값만으로 모델의 성능을 평가하는 것은 불충분하며, 다른 평가 지표들과 함께 고려해야 합니다.

머신러닝 모델은 여러 가지 방법으로 예측 성능을 평가할 수 있습니다. 성능 평가 지표(Evaluation Metrics)는 일반적으로 모델이 분류냐 회귀냐에 따라 여러 종류로 나뉩니다.

회귀의 경우 대부분 실제값과 예측값의 오차 평균값에 기반합니다. 예를 들어 오차에 절댓값을 씌운 뒤 평균 오차를 구하거나 오차의 제곱 값에 루트를 씌운 뒤 평균 오차를 구하는 방법과 같이 기본적으로 예측 오차를 가지고 정규화 수준을 재가공하는 방법이 회귀의 성능 평가 지표 유형입니다.

분류의 평가방법도 일반적으로는 실제 결과 데이터와 예측 결과 데이터가 얼마나 정확하고 오류가 적게 발생하는가에 기반하지만, 단순히 이러한 정확도만 가지고 판단했다가는 잘못된 평가 결과에 빠질수 있습니다. 특히 0과 1로 결정값이 한정되는 이진 분류의 성능 평가 지표는 정확도보다는 다른 성능 평가 지표가 더 중요시되는 경우가 많습니다.

﻿

**Comment by Romangduck** [comment]
https://blog.naver.com/rrlagusejr/223097632282
